File: APIClient.Tests/ServicesTests/Constructor/When_preLoadMeta_is_false.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace VersionOne.SDK.APIClient.Tests.ServicesTests.Constructor
{
    using static ServicesConstructorTestsHelpers;

    [TestClass]
    public class When_preLoadMeta_is_false
    {
        private static IServices SUT;
        private static IAssetType AssetTypeType;
        private static IAssetType PrimaryRelationType;
        private static TestContext Context;

        [ClassInitialize]
        public static void Initialize(TestContext context)
        {
            Context = context;
            Configure(Context);
            ConfigureRoute(Context, "/meta.v1//AssetType", MetaSamplePayloads.AssetTypeType);
            ConfigureRoute(Context, "/meta.v1//PrimaryRelation", MetaSamplePayloads.PrimaryRelationType);

            var connector = CreateConnector();
            SUT = new Services(connector);

            AssetTypeType = SUT.Meta.GetAssetType("AssetType");
            PrimaryRelationType = SUT.Meta.GetAssetType("PrimaryRelation");
        }

        [TestMethod]
        public void It_should_not_call_full_meta_route() => AssertRouteNotCalled(Context, "/meta.v1//");

        [TestMethod]
        public void It_should_let_me_get_the_AssetType_type() => Assert.IsNotNull(AssetTypeType);

        [TestMethod]
        public void It_should_access_the_AssetType_route() => AssertRouteCalled(Context, "/meta.v1//AssetType");

        [TestMethod]
        public void It_should_let_me_get_the_PrimaryRelation_type() => Assert.IsNotNull(PrimaryRelationType);

        [TestMethod]
        public void It_should_accecss_the_PrimaryRelation_route() => AssertRouteCalled(Context, "/meta.v1//PrimaryRelation");
    }
}

File: APIClient.FSpec.Tests/When_preLoadMeta_is_false.fs
module When_preLoadMeta_is_false
open FSpec.Dsl
open FSpec.Matchers
open VersionOne.SDK.APIClient
open ServicesConstructorHelpers
open MetaSamplePayloads

let specs =
    describe "Services" [
        describe "Constructor" [
            context "When preLoadMeta is false" [
                subject <| fun ctx ->
                    configure(ctx)
                    configureRoute ctx "/meta.v1//AssetType" assetTypeType
                    configureRoute ctx "/meta.v1//PrimaryRelation" primaryRelationType

                    let connector = createConnector
                    Services(connector)
                
                before <| fun ctx ->
                    let subject = ctx.GetSubject<IServices>()
                    ctx?assetTypeType <- subject.Meta.GetAssetType "AssetType"
                    ctx?primaryRelationType <- subject.Meta.GetAssetType "PrimaryRelation"

                it "should not call the full meta.v1 route" <| fun ctx -> assertRouteNotCalled ctx "meta.v1//"
                
                it "should access the AssetType route" <| fun ctx -> assertRouteCalled ctx "/meta.v1//AssetType"
                it "should let me access the AssetType type" <| fun ctx -> ctx?assetTypeType |> shouldNot (equal null)

                it "should access the AssetType route" <| fun ctx -> assertRouteCalled ctx "/meta.v1//PrimaryRelation"
                it "should let me access the PrimaryRelation type" <| fun ctx -> ctx?primaryRelationType |> shouldNot (equal null)
            ]]]